### ISSUE TITLE
Support pluggable ethernet devices (avoid "Timed out while waiting for device")

### ISSUE THREAD

**Issue by rien** [NixOS/nixpkgs#154737]
### Describe the bug

I have configured a pluggable ethernet device (from a dock) on my NixOS system with `networking.interfaces.enp0s13f0u1u4.useDHCP = true`. When I boot without the dock plugged in, it now waits 1m30s for the ethernet device to appear (which won't happen) and then proceed booting normally. 

### Steps To Reproduce
1. Configure a pluggable or non-existing ethernet device
2. Boot without the device 
3. Systemd will show a message: `Waiting for a start job for device ...`
4. After 1m30s, systemd will time out logging `Timed out waiting for device /sys/subsystem/net/devices/enp0s13f0u1u4`
5. The boot completes succesfully

### Expected behavior

If the pluggable ethernet device is not present when booting, systemd should not needlessly delay the booting process by waiting on the device to appear.

This should probably not be the default behavior however, since some systems depend on an ethernet connection to be present. We probably want an option to set `networking.interfaces.<name>.pluggable = true;` to enable skipping the device at boot time.

### Additional context
I am using a Dell XPS 13 with a dock.

### Notify maintainers

@symphorien, @Artturin, @rnhmjoj

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.10, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4`
 - channels(rien): `""`
 - nixpkgs: `/etc/nixpkgs`
```

**Comment by rnhmjoj** [comment]
If don't think the `networking.interfaces` or `networking.dhcpcd` modules are  setting a dependecy on the network devices: if they do I can't find where in Nixpkgs. I guess you've configured some other software that does (`networking.supplicant` or `networking.wireless`, maybe?).
Can you tell me what does this say?
```
systemctl list-dependencies --reverse sys-subsystem-net-devices-enp0s13f0u1u4.device 
```

**Comment by rien** [comment]
```
[rien@chaos]$ systemctl list-dependencies --reverse sys-subsystem-net-devices-enp0s13f0u1u4.device
sys-subsystem-net-devices-enp0s13f0u1u4.device
○ └─network-addresses-enp0s13f0u1u4.service
```
Which in turn is a dependency of `network-setup.service` and part of the `network.target`:
```
[rien@chaos]$ systemctl list-dependencies --reverse network-addresses-enp0s13f0u1u4.service
network-addresses-enp0s13f0u1u4.service
● ├─network-setup.service
● └─network.target
○   ├─cups.service
●   ├─dhcpcd.service
●   ├─network-setup.service
●   └─wpa_supplicant-wlp114s0.service
```
Which seems to be defined at https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/tasks/network-interfaces-scripted.nix#L170

You should be able to reproduce this locally by adding `networking.interfaces.idontexist.useDHCP = true;` to your NixOS config.

**Comment by rnhmjoj** [comment]
I see, thank you. Yes, I think we could define an option `interfaces.<name>.neededForBoot`, like the one for disks, and
do something like:
```
bindsTo = optional i.neededForBoot (deviceDependency i.name);
```

**Comment by dezren39** [comment]
this issue persists. is there something that can be done in the meantime?

**Comment by rnhmjoj** [comment]
No, you can try to resume the work in https://github.com/NixOS/nixpkgs/pull/155017 if you're interested.

**Comment by dezren39** [comment]
I've attempted to resurrect #155017 in #360092. Instead of removing the `bindsTo`, I removed the `wantedBy`. I also incorporated the requested change around the option `description`. We can continue discussion there as-needed, so far it _appears_ to work on my machine -- although I haven't done thorough testing.

**Comment by fidgetingbits** [comment]
I've recently set up two beelink mini pc's with built-in wireless cards using [facter](https://github.com/nix-community/nixos-facter) and noticed that I run into this wait time for them without any explicit wlo1 configuration. I've worked around it by just removing the wireless entries from facter.json for now, as I don't need wireless, but I guess this feature will help here as well. Just noting in case anyone else has run into it from this angle yet. I'll hopefully have time to try #360092 to see if it addresses it.

**Comment by rnhmjoj** [comment]
I'll try to resume that PR as soon as I have some spare time. It shouldn't be too much work.

**Comment by rnhmjoj** [comment]
Please check #481473 for a fix.
